source/game/kart/KartPullPath.cc

```diff
@@ -0,0 +1,286 @@
+    const auto &nextPos = points[m_currentIdx + n].pos;
+
+    EGG::Vector3f back = currentPos - nextPos;
+    back.normalise();
+    EGG::Vector3f front = -back;
+
+    if (m_currentIdx + p >= 0 && static_cast<size_t>(m_currentIdx + p) < points.size()) {
+        const auto &prevPos = points[m_currentIdx + p].pos;
+        back = prevPos - currentPos;
+        back.normalise();
+    }
+
+    EGG::Plane3f backPlane = EGG::Plane3f(currentPos, back);
+    EGG::Plane3f frontPlane = EGG::Plane3f(nextPos, front);
+
+    if (backPlane.testPoint(pos()) && frontPlane.testPoint(pos())) {
+        point = currentPos;
+        dir = front;
+        return true;
+    }
+
+    return false;
+}
+
+/// @addr{0x80593FA4}
+KartPullPath::KartPullPath()
+    : m_globalTracker(this, KartPullPathTracker::Type::Global),
+      m_regionalTracker(this, KartPullPathTracker::Type::Regional) {
+    init();
+}
+
+/// @addr{0x80594094}
+KartPullPath::~KartPullPath() = default;
+
+/// @addr{0x805940D4}
+void KartPullPath::init() {
+    reset();
+    m_areaId = -1;
+    m_roadSpeedDecay = 1.0f;
+}
+
+/// @addr{0x80593CB8}
+/// @details This was the init function in the base class, but it gets inlined in calcArea.
+void KartPullPath::reset() {
+    m_distance = -1.0f;
+    m_currentIdx = -1;
+    m_incomingIdx = -1;
+    m_pointInfo = nullptr;
+    m_pullDirection = EGG::Vector3f::zero;
+    m_pullSpeed = 0.0f;
+    m_maxPullSpeed = 0.0f;
+}
+
+/// @addr{0x80594134}
+void KartPullPath::calc() {
+    if (!calcArea()) {
+        return;
+    }
+
+    ASSERT(m_pointInfo);
+    if (m_pointInfo->pointCount() > 2) {
+        calcTrackers();
+    }
+}
+
+/// @addr{0x80593DBC}
+void KartPullPath::changePoint(s16 idx, f32 distance) {
+    if ((m_distance >= 0.0f || distance >= 3000.0f) && distance >= m_distance) {
+        return;
+    }
+
+    m_incomingIdx = idx;
+    m_distance = distance;
+    m_regionalTracker.setCurrentIdx(idx);
+    calcPointChange();
+}
+
+/// @addr{0x805941BC}
+bool KartPullPath::calcArea() {
+    auto *courseMap = System::CourseMap::Instance();
+
+    s16 prevAreaId = m_areaId;
+    m_areaId = System::CourseMap::Instance()->getCurrentAreaID(m_areaId, pos(),
+            System::MapdataAreaBase::Type::MovingRoad);
+
+    if (m_areaId >= 0 && (prevAreaId < 0 || prevAreaId != m_areaId)) {
+        reset();
+
+        auto *area = courseMap->getArea(m_areaId);
+        ASSERT(area);
+        auto *pointInfo = area->getPointInfo();
+        ASSERT(pointInfo);
+        m_pointInfo = pointInfo;
+
+        if (pointInfo->pointCount() > 2) {
+            setTrackerPointInfo(pointInfo);
+        } else {
+            m_incomingIdx = 0;
+            calcPointChange();
+        }
+
+        m_roadSpeedDecay = 0.9f + 0.001f * static_cast<f32>(area->param(0));
+        m_maxPullSpeed = static_cast<f32>(area->param(1));
+    } else {
+        m_areaId = -1;
+        m_currentIdx = -1;
+        m_incomingIdx = -1;
+        m_pullDirection = EGG::Vector3f::zero;
+    }
```

Comment:
Not equivalent to the base game
```suggestion
    if (m_areaId >= 0) {
        if (prevAreaId < 0 || prevAreaId != m_areaId) {
            reset();

            auto *area = courseMap->getArea(m_areaId);
            ASSERT(area);
            auto *pointInfo = area->getPointInfo();
            ASSERT(pointInfo);
            m_pointInfo = pointInfo;

            if (pointInfo->pointCount() > 2) {
                setTrackerPointInfo(pointInfo);
            } else {
                m_incomingIdx = 0;
                calcPointChange();
            }

            m_roadSpeedDecay = 0.9f + 0.001f * static_cast<f32>(area->param(0));
            m_maxPullSpeed = static_cast<f32>(area->param(1));
        }
    } else {
        m_areaId = -1;
        m_currentIdx = -1;
        m_incomingIdx = -1;
        m_pullDirection = EGG::Vector3f::zero;
    }
```